File: main/src/main/java/com/google/android/apps/muzei/render/DemoRenderController.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.render;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.os.Handler;
import android.util.Log;

import java.io.IOException;

public class DemoRenderController extends RenderController {
    private static final String TAG = "DemoRenderController";

    private final Handler mHandler = new Handler();

    private static final long ANIMATION_CYCLE_TIME_MILLIS = 35000;
    private static final long FOCUS_DELAY_TIME_MILLIS = 2000;
    private static final long FOCUS_TIME_MILLIS = 6000;

    private Animator mCurrentScrollAnimator;
    private boolean mReverseDirection = false;
    private boolean mAllowFocus = true;

    public DemoRenderController(Context context, MuzeiBlurRenderer renderer,
            Callbacks callbacks, boolean allowFocus) {
        super(context, renderer, callbacks);
        mAllowFocus = allowFocus;
        runAnimation();
    }

    private void runAnimation() {
        if (mCurrentScrollAnimator != null) {
            mCurrentScrollAnimator.cancel();
        }

        mCurrentScrollAnimator = ObjectAnimator
                .ofFloat(mRenderer, "normalOffsetX",
                        mReverseDirection ? 1f : 0f, mReverseDirection ? 0f : 1f)
                .setDuration(ANIMATION_CYCLE_TIME_MILLIS);
        mCurrentScrollAnimator.start();
        mCurrentScrollAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                mReverseDirection = !mReverseDirection;
                runAnimation();
            }
        });
        if (mAllowFocus) {
            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    mRenderer.setIsBlurred(false, false);
                    mHandler.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            mRenderer.setIsBlurred(true, false);
                        }
                    }, FOCUS_TIME_MILLIS);
                }
            }, FOCUS_DELAY_TIME_MILLIS);
        }
    }

    @Override
    public void destroy() {
        super.destroy();
        if (mCurrentScrollAnimator != null) {
            mCurrentScrollAnimator.cancel();
        }
        mHandler.removeCallbacksAndMessages(null);
    }

    @Override
    protected BitmapRegionLoader openDownloadedCurrentArtwork(boolean forceReload) {
        try {
            return BitmapRegionLoader.newInstance(mContext.getAssets().open("starrynight.jpg"));
        } catch (IOException e) {
            Log.e(TAG, "Error opening demo image.", e);
            return null;
        }
    }
}

File: main/src/main/java/com/google/android/apps/muzei/render/DemoRenderController.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.render

import android.animation.Animator
import android.animation.ObjectAnimator
import android.content.Context
import androidx.core.animation.doOnEnd
import androidx.lifecycle.LifecycleOwner
import com.google.android.apps.muzei.util.coroutineScope
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class DemoRenderController(
        context: Context,
        renderer: MuzeiBlurRenderer,
        callbacks: RenderController.Callbacks,
        private val allowFocus: Boolean
) : RenderController(context, renderer, callbacks) {

    companion object {
        private const val ANIMATION_CYCLE_TIME_MILLIS = 35000L
        private const val FOCUS_DELAY_TIME_MILLIS = 2000L
        private const val FOCUS_TIME_MILLIS = 6000L
    }

    private lateinit var coroutineScope: CoroutineScope

    private var currentScrollAnimator: Animator? = null
    private var reverseDirection = false

    private fun runAnimation() {
        currentScrollAnimator?.cancel()

        currentScrollAnimator = ObjectAnimator
                .ofFloat(renderer, "normalOffsetX",
                        if (reverseDirection) 1f else 0f, if (reverseDirection) 0f else 1f).apply {

                    duration = ANIMATION_CYCLE_TIME_MILLIS
                    doOnEnd {
                        reverseDirection = !reverseDirection
                        runAnimation()
                    }
                    start()
                }
        if (allowFocus) {
            coroutineScope.launch {
                delay(FOCUS_DELAY_TIME_MILLIS)
                renderer.setIsBlurred(false, false)
                delay(FOCUS_TIME_MILLIS)
                renderer.setIsBlurred(true, false)
            }
        }
    }

    override fun onCreate(owner: LifecycleOwner) {
        super.onCreate(owner)
        coroutineScope = owner.coroutineScope
        runAnimation()
    }

    override fun onDestroy(owner: LifecycleOwner) {
        super.onDestroy(owner)
        currentScrollAnimator?.cancel()
        currentScrollAnimator?.removeAllListeners()
    }

    override suspend fun openDownloadedCurrentArtwork() =
            AssetImageLoader(context.assets, "starrynight.jpg")
}
